APE_public.rst

```diff
@@ -0,0 +1,284 @@
+an ``__all__`` attribute in every module is simpler, unambiguous, and better for
+introspection by both users and automated systems.
+
+Finally, let's consider what `Python typing guide
+<https://github.com/python/typing/blob/master/docs/source/libraries.rst#library-interface-public-and-private-symbols>`
+adds. :: 
+
+   - Symbols whose names begin with an underscore (but are not dunder names) are
+     considered private.
+
+   - Imported symbols are considered private by default. If they use the
+     ``import A as A`` (a redundant module alias), from ``X import A as A``
+     (a redundant symbol alias), or ``from . import A`` forms, symbol ``A`` is
+     not private unless the name begins with an underscore. If a file
+     ``__init__.py`` uses form ``from .A import X``, symbol ``A`` is treated
+     likewise. If a wildcard import (of the form ``from X import *``) is used,
+     all symbols referenced by the wildcard are not private.
+   
+   - A module can expose an ``__all__`` symbol at the module level that provides
+     a list of names that are considered part of the interface. This overrides
+     all other rules above, allowing imported symbols or symbols whose names
+     begin with an underscore to be included in the interface.
+   
+   - Local variables within a function (including nested functions) are always
+     considered private.
+
+This is consistent with PEP 8 and SciPy, but it does not mention the
+documentation. This typing recommendations reinforces the idea that the
+``__all__`` attribute is the authoritative source of what is public vs internal.
+Importantly, the ``__all__`` attribute is a module-level attribute, so it only
+applies to the module in which it is defined.  This means that a module can
+define an ``__all__`` attribute but if the module itself is not public, then
+anything in the ``__all__`` attribute cannot be publicly accessed from outside
+the module. For example, consider the following module::
+
+   src/
+      __init__/
+         __all__ = []
+
+      _foo/
+         __all__ = ['bar']
+         def bar():
+            pass
+
+In this case, ``bar`` is public in ``_foo``, but  ``src._foo`` is not public, so
+``src._foo.bar`` is not public either.
+
+We propose the following:
+
+1. That Astropy adopts all the PEP 8 rules on public versus internal interfaces
+   (the only one I'm not sure about is adding prefixes to private classes and
+   functions.)
+2. That Astropy disambiguates these rules by adopting the subsequent rules for
+   its API:
+3. That Astropy ensures its documentation is consistent with its code, the
+   latter being the authoritative source.
+
+   - That Astropy explains these rules in its documentation
+4. That Astropy also adopts these rules for coordinated packages, and recommends
+   that affiliated packages follow these rules as well.
+
+Rules:
+
+1. A symbol is public if all containing namespaces (packages, modules, classes,
```

Comment:
Is that *if and only if*?

Reply:
Yes. A connected graph through Public API. No private islands for non-dev users.